bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -220,53 +230,90 @@ protected void checkedStart(Listener<RespT> responseListener, Metadata headers)
    *
    * @throws IOException
    */
-  public void asyncRefresh() throws IOException {
+  ListenableFuture<HeaderCacheElement> asyncRefresh() {
     if (isAppEngine) {
-      syncRefresh();
-    } else if (canRefresh()) {
-      executor.execute(new Runnable() {
-        @Override
-        public void run() {
-          doRefresh();
-        }
-      });
+      SettableFuture<HeaderCacheElement> f = SettableFuture.create();
+      try {
+        f.set(syncRefresh());
+      } catch (Exception e) {
+        f.setException(e);
+      }
+      return f;
     }
-  }
 
-  private boolean canRefresh() {
-    return !isRefreshing.get() && getCacheState(this.headerCache.get()) != CacheState.Good;
+    synchronized (this.isRefreshing) {
+      if (!this.isRefreshing.compareAndSet(false, true)) {
+        this.futureToken = executor.submit(new Callable<HeaderCacheElement>() {
+          @Override
+          public HeaderCacheElement call() throws Exception {
+            HeaderCacheElement newToken = refreshCredentialsWithRetry();
```

Comment:
I get it now :).  The `readerFuture.get(250, TimeUnit.MILLISECONDS); ` will deliver the "notification"